File: lib/gis/direction-mode.ts
/**
 * Corredor canónico + sentidos ida/vuelta.
 *
 * - mirrored: una geometría canónica (ida); vuelta = reverse(ida).
 * - independent: ida y vuelta tienen trazos distintos (one-ways, desvíos).
 *
 * Fase 1: UI muestra un sentido a la vez; mirror en editor; flag en GeoJSON.
 * Fase 2 (futuro): un solo LineString canónico en storage.
 */

export type RouteDirection = 'ida' | 'vuelta';
export type DirectionMode = 'mirrored' | 'independent' | 'dual_ring';

export type LngLat = [number, number];

export interface RouteFeatureProperties {
  direction?: string;
  name?: string;
  directionMode?: DirectionMode;
  [key: string]: unknown;
}

export interface RouteFeature {
  type: 'Feature' | string;
  properties?: RouteFeatureProperties | null;
  geometry?: {
    type?: string;
    coordinates?: LngLat[] | number[][];
  } | null;
}

export interface RouteFeatureCollection {
  type: 'FeatureCollection' | string;
  features?: RouteFeature[];
  properties?: Record<string, unknown>;
  [key: string]: unknown;
}

/** Umbral por defecto (m): media de distancias ida ↔ reverse(vuelta) para considerar mirrored. */
export const DEFAULT_MIRROR_THRESHOLD_M = 25;

export function directionOfFeature(f: RouteFeature | null | undefined): string {
  return String(f?.properties?.direction ?? f?.properties?.name ?? '').toLowerCase();
}

export function getDirectionMode(
  geojson: RouteFeatureCollection | null | undefined
): DirectionMode {
  const fromProps = geojson?.properties?.directionMode;
  if (fromProps === 'mirrored' || fromProps === 'independent') return fromProps;

  const feat = (geojson?.features ?? []).find(
    (f) => f?.properties?.directionMode === 'mirrored' || f?.properties?.directionMode === 'independent'
  );
  if (feat?.properties?.directionMode === 'mirrored') return 'mirrored';
  if (feat?.properties?.directionMode === 'independent') return 'independent';

  return 'independent';
}

export function reverseCoordinates(coords: LngLat[]): LngLat[] {
  return coords.map((c) => [c[0], c[1]] as LngLat).reverse();
}

export function haversineMeters(a: LngLat, b: LngLat): number {
  const R = 6371000;
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(b[1] - a[1]);
  const dLon = toRad(b[0] - a[0]);
  const lat1 = toRad(a[1]);
  const lat2 = toRad(b[1]);
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLon / 2) ** 2;
  return 2 * R * Math.asin(Math.min(1, Math.sqrt(h)));
}

function sampleIndices(n: number, maxSamples: number): number[] {
  if (n <= 0) return [];
  if (n <= maxSamples) return Array.from({ length: n }, (_, i) => i);
  const out: number[] = [];
  for (let i = 0; i < maxSamples; i++) {
    out.push(Math.round((i * (n - 1)) / (maxSamples - 1)));
  }
  return out;
}

/** Distancia media (m) de puntos de A al más cercano en B (muestreado). */
export function meanNearestDistanceMeters(
  a: LngLat[],
  b: LngLat[],
  maxSamples = 40
): number {
  if (a.length === 0 || b.length === 0) return Infinity;
  const aIdx = sampleIndices(a.length, maxSamples);
  let sum = 0;
  for (const i of aIdx) {
    const pt = a[i];
    let best = Infinity;
    // subsample B for speed
    const bStep = Math.max(1, Math.floor(b.length / 80));
    for (let j = 0; j < b.length; j += bStep) {
      const d = haversineMeters(pt, b[j]);
      if (d < best) best = d;
    }
    // always check endpoints of B
    best = Math.min(best, haversineMeters(pt, b[0]), haversineMeters(pt, b[b.length - 1]));
    sum += best;
  }
  return sum / aIdx.length;
}

/**
 * Métrica simétrica: ida vs reverse(vuelta).
 * Si es baja, los trazos son el mismo corredor en sentidos opuestos.
 */
export function mirrorSimilarityMeters(
  ida: LngLat[],
  vuelta: LngLat[]
): number {
  if (ida.length < 2 || vuelta.length < 2) return Infinity;
  const revVuelta = reverseCoordinates(vuelta);
  const d1 = meanNearestDistanceMeters(ida, revVuelta);
  const d2 = meanNearestDistanceMeters(revVuelta, ida);
  return (d1 + d2) / 2;
}

export function detectDirectionMode(
  ida: LngLat[],
  vuelta: LngLat[],
  thresholdM = DEFAULT_MIRROR_THRESHOLD_M
): { mode: DirectionMode; similarityM: number } {
  const similarityM = mirrorSimilarityMeters(ida, vuelta);
  return {
    mode: similarityM <= thresholdM ? 'mirrored' : 'independent',
    similarityM,
  };
}

export function findDirectionFeature(
  geojson: RouteFeatureCollection | null | undefined,
  direction: RouteDirection
): RouteFeature | undefined {
  return (geojson?.features ?? []).find((f) => directionOfFeature(f) === direction);
}

export function getDirectionCoords(
  geojson: RouteFeatureCollection | null | undefined,
  direction: RouteDirection
): LngLat[] {
  const f = findDirectionFeature(geojson, direction);
  const raw = f?.geometry?.coordinates;
  if (!Array.isArray(raw)) return [];
  return raw.map((c) => [Number(c[0]), Number(c[1])] as LngLat);
}

/** GeoJSON de preview/mapa: solo un sentido (evita línea doble). */
export function filterGeojsonByDirection(
  geojson: RouteFeatureCollection,
  direction: RouteDirection,
  draftCoords?: LngLat[]
): RouteFeatureCollection {
  const features = (geojson.features ?? [])
    .filter((f) => directionOfFeature(f) === direction)
    .map((f) => {
      if (!draftCoords) return { ...f };
      return {
        ...f,
        geometry: {
          ...f.geometry,
          type: 'LineString',
          coordinates: draftCoords,
        },
      };
    });

  return {
    type: 'FeatureCollection',
    features,
  };
}

/**
 * Prepara GeoJSON de visualización de una ruta.
 *
 * - dual_ring / both_kml_lines: dibuja LAS DOS líneas reales del KML (como el PDF),
 *   misma color, etiquetas Ida y Vuelta en cada sentido real.
 * - mirrored (vuelta≈reverse(ida)): una sola geometría + sense-labels.
 * - independent: ambas líneas reales (como dual_ring).
 */
export function toSingleCorridorDisplay(
  geojson: RouteFeatureCollection,
  options?: {
    color?: string;
    casingColor?: string;
    preferDirection?: RouteDirection;
    role?: string;
  }
): RouteFeatureCollection {
  const ida = findDirectionFeature(geojson, 'ida');
  const vuelta = findDirectionFeature(geojson, 'vuelta');
  const idaCoords = getDirectionCoords(geojson, 'ida');
  const vueltaCoords = getDirectionCoords(geojson, 'vuelta');
  const mode =
    String(
      geojson.properties?.directionMode ??
        ida?.properties?.directionMode ??
        vuelta?.properties?.directionMode ??
        geojson.properties?.corridor ??
        ''
    ).toLowerCase();
  const corridor = String(
    geojson.properties?.corridor ?? ida?.properties?.corridor ?? ''
  ).toLowerCase();

  const color =
    options?.color ||
    String(ida?.properties?.color || vuelta?.properties?.color || '#3b82f6');
  const casingColor =
    options?.casingColor ||
    String(ida?.properties?.casingColor || vuelta?.properties?.casingColor || '#222222');
  const role = options?.role ?? 'full';

  // PDF dual (anillo ida+vuelta reales): dibujar ambas
  const drawBoth =
    mode === 'dual_ring' ||
    mode === 'independent' ||
    corridor === 'both_kml_lines' ||
    corridor === 'dual_ring' ||
    (idaCoords.length >= 2 &&
      vueltaCoords.length >= 2 &&
      // si no es reverse exacto, son dos trazos reales
      JSON.stringify(vueltaCoords) !== JSON.stringify(reverseCoordinates(idaCoords)));

  if (drawBoth && idaCoords.length >= 2 && vueltaCoords.length >= 2) {
    return {
      type: 'FeatureCollection',
      properties: {
        ...(geojson.properties ?? {}),
        directionMode: 'dual_ring',
        corridor: 'both_kml_lines',
      },
      features: [
        {
          type: 'Feature',
          properties: {
            ...(ida?.properties ?? {}),
            type: 'ride',
            role,
            direction: 'ida',
            name: 'Ida',
            color,
            casingColor,
            directionMode: 'dual_ring',
          },
          geometry: { type: 'LineString', coordinates: idaCoords },
        },
        {
          type: 'Feature',
          properties: {
            ...(vuelta?.properties ?? {}),
            type: 'ride',
            role,
            direction: 'vuelta',
            name: 'Vuelta',
            color,
            casingColor,
            directionMode: 'dual_ring',
          },
          geometry: { type: 'LineString', coordinates: vueltaCoords },
        },
      ],
    };
  }

  // Corredor único (vuelta = reverse): 1 línea + etiquetas de sentido
  let corridorCoords: LngLat[] = [];
  let baseProps: RouteFeatureProperties = {};

  const prefer = options?.preferDirection;
  if (prefer === 'vuelta' && vueltaCoords.length >= 2) {
    corridorCoords = vueltaCoords;
    baseProps = { ...(vuelta?.properties ?? {}) };
  } else if (idaCoords.length >= 2) {
    corridorCoords = idaCoords;
    baseProps = { ...(ida?.properties ?? {}) };
  } else if (vueltaCoords.length >= 2) {
    corridorCoords = vueltaCoords;
    baseProps = { ...(vuelta?.properties ?? {}) };
  } else {
    const first = (geojson.features ?? []).find(
      (f) => f.geometry?.type === 'LineString' && (f.geometry.coordinates?.length ?? 0) >= 2
    );
    if (first?.geometry?.coordinates) {
      corridorCoords = first.geometry.coordinates.map(
        (c) => [Number(c[0]), Number(c[1])] as LngLat
      );
      baseProps = { ...(first.properties ?? {}) };
    }
  }

  if (corridorCoords.length < 2) {
    return { type: 'FeatureCollection', features: [] };
  }

  const rev = reverseCoordinates(corridorCoords);

  return {
    type: 'FeatureCollection',
    properties: {
      ...(geojson.properties ?? {}),
      directionMode: 'mirrored',
    },
    features: [
      {
        type: 'Feature',
        properties: {
          ...baseProps,
          type: 'ride',
          role,
          directionMode: 'mirrored',
          color,
          casingColor,
          name: '',
          direction: 'corridor',
        },
        geometry: { type: 'LineString', coordinates: corridorCoords },
      },
      {
        type: 'Feature',
        properties: {
          type: 'sense-label',
          name: 'Ida',
          direction: 'ida',
          color,
          casingColor,
          directionMode: 'mirrored',
        },
        geometry: { type: 'LineString', coordinates: corridorCoords },
      },
      {
        type: 'Feature',
        properties: {
          type: 'sense-label',
          name: 'Vuelta',
          direction: 'vuelta',
          color,
          casingColor,
          directionMode: 'mirrored',
        },
        geometry: { type: 'LineString', coordinates: rev },
      },
    ],
  };
}

/** Aplica directionMode a collection + cada feature. */
export function stampDirectionMode(
  geojson: RouteFeatureCollection,
  mode: DirectionMode
): RouteFeatureCollection {
  const next: RouteFeatureCollection = {
    ...geojson,
    type: 'FeatureCollection',
    properties: {
      ...(geojson.properties ?? {}),
      directionMode: mode,
    },
    features: (geojson.features ?? []).map((f) => ({
      ...f,
      properties: {
        ...(f.properties ?? {}),
        directionMode: mode,
      },
    })),
  };
  return next;
}

/**
 * Copia coordenadas de `from` a `to` como reverse.
 * Crea feature de destino si no existe (clonando props del origen).
 */
export function mirrorDirection(
  geojson: RouteFeatureCollection,
  from: RouteDirection,
  to: RouteDirection
): RouteFeatureCollection {
  const next = JSON.parse(JSON.stringify(geojson)) as RouteFeatureCollection;
  const source = findDirectionFeature(next, from);
  if (!source?.geometry?.coordinates || source.geometry.coordinates.length < 2) {
    return stampDirectionMode(next, 'mirrored');
  }

  const coords = reverseCoordinates(
    source.geometry.coordinates.map((c) => [Number(c[0]), Number(c[1])] as LngLat)
  );

  let targetIdx = (next.features ?? []).findIndex((f) => directionOfFeature(f) === to);
  if (targetIdx < 0) {
    const clone: RouteFeature = JSON.parse(JSON.stringify(source));
    clone.properties = {
      ...(clone.properties ?? {}),
      direction: to,
      name: to === 'ida' ? 'Ida' : 'Vuelta',
    };
    clone.geometry = { type: 'LineString', coordinates: coords };
    next.features = [...(next.features ?? []), clone];
  } else {
    const t = next.features![targetIdx];
    t.geometry = {
      ...(t.geometry ?? {}),
      type: 'LineString',
      coordinates: coords,
    };
    t.properties = {
      ...(t.properties ?? {}),
      direction: to,
      name: to === 'ida' ? 'Ida' : 'Vuelta',
    };
  }

  return stampDirectionMode(next, 'mirrored');
}

/**
 * Tras editar un sentido en modo mirrored, sincroniza el opuesto = reverse(editado).
 * En independent solo actualiza el sentido editado.
 */
export function applyEditedDirection(
  geojson: RouteFeatureCollection,
  direction: RouteDirection,
  coords: LngLat[],
  options?: {
    /** true: siempre reverse al otro; false: nunca; omitido: si mode=mirrored */
    forceMirrorSync?: boolean;
  }
): RouteFeatureCollection {
  const next = JSON.parse(JSON.stringify(geojson)) as RouteFeatureCollection;
  const mode = getDirectionMode(next);
  // forceMirrorSync: true → siempre; false → nunca; undefined → si mode=mirrored
  const syncMirror =
    options?.forceMirrorSync === true ||
    (options?.forceMirrorSync !== false && mode === 'mirrored');

  let idx = (next.features ?? []).findIndex((f) => directionOfFeature(f) === direction);
  if (idx < 0) {
    // crear feature mínima
    next.features = next.features ?? [];
    next.features.push({
      type: 'Feature',
      properties: {
        direction,
        name: direction === 'ida' ? 'Ida' : 'Vuelta',
        directionMode: syncMirror ? 'mirrored' : mode,
      },
      geometry: { type: 'LineString', coordinates: coords },
    });
  } else {
    next.features![idx].geometry = {
      ...(next.features![idx].geometry ?? {}),
      type: 'LineString',
      coordinates: coords,
    };
    next.features![idx].properties = {
      ...(next.features![idx].properties ?? {}),
      direction,
      name: direction === 'ida' ? 'Ida' : 'Vuelta',
    };
  }

  if (syncMirror && coords.length >= 2) {
    const other: RouteDirection = direction === 'ida' ? 'vuelta' : 'ida';
    return mirrorDirection(next, direction, other);
  }

  return stampDirectionMode(next, mode);
}

/** Detecta y sella directionMode en un FeatureCollection con ida+vuelta. */
export function annotateDirectionMode(
  geojson: RouteFeatureCollection,
  thresholdM = DEFAULT_MIRROR_THRESHOLD_M
): { geojson: RouteFeatureCollection; mode: DirectionMode; similarityM: number } {
  const ida = getDirectionCoords(geojson, 'ida');
  const vuelta = getDirectionCoords(geojson, 'vuelta');
  const { mode, similarityM } = detectDirectionMode(ida, vuelta, thresholdM);
  return {
    geojson: stampDirectionMode(geojson, mode),
    mode,
    similarityM,
  };
}

File: viamoreliaapp/lib/gis/direction_mode.dart
import 'dart:math';
import 'package:latlong2/latlong.dart';
import '../models/route_model.dart';

/// Port de `lib/gis/direction-mode.ts` (web).
enum DirectionMode { mirrored, independent, dualRing }

class DirectionModeService {
  static const double defaultMirrorThresholdM = 25;

  static List<LatLng> reverseCoordinates(List<LatLng> coords) =>
      coords.reversed.map((c) => LatLng(c.latitude, c.longitude)).toList();

  static double haversineMeters(LatLng a, LatLng b) {
    const r = 6371000.0;
    final dLat = _rad(b.latitude - a.latitude);
    final dLon = _rad(b.longitude - a.longitude);
    final lat1 = _rad(a.latitude);
    final lat2 = _rad(b.latitude);
    final h = sin(dLat / 2) * sin(dLat / 2) +
        cos(lat1) * cos(lat2) * sin(dLon / 2) * sin(dLon / 2);
    return 2 * r * asin(min(1, sqrt(h)));
  }

  static double _rad(double d) => d * pi / 180;

  static List<int> _sampleIndices(int n, int maxSamples) {
    if (n <= 0) return [];
    if (n <= maxSamples) return List.generate(n, (i) => i);
    return List.generate(
      maxSamples,
      (i) => ((i * (n - 1)) / (maxSamples - 1)).round(),
    );
  }

  static double meanNearestDistanceMeters(
    List<LatLng> a,
    List<LatLng> b, {
    int maxSamples = 40,
  }) {
    if (a.isEmpty || b.isEmpty) return double.infinity;
    final aIdx = _sampleIndices(a.length, maxSamples);
    var sum = 0.0;
    final bStep = max(1, b.length ~/ 80);
    for (final i in aIdx) {
      final pt = a[i];
      var best = double.infinity;
      for (var j = 0; j < b.length; j += bStep) {
        final d = haversineMeters(pt, b[j]);
        if (d < best) best = d;
      }
      best = min(best, haversineMeters(pt, b.first));
      best = min(best, haversineMeters(pt, b.last));
      sum += best;
    }
    return sum / aIdx.length;
  }

  static double mirrorSimilarityMeters(List<LatLng> ida, List<LatLng> vuelta) {
    if (ida.length < 2 || vuelta.length < 2) return double.infinity;
    final rev = reverseCoordinates(vuelta);
    final d1 = meanNearestDistanceMeters(ida, rev);
    final d2 = meanNearestDistanceMeters(rev, ida);
    return (d1 + d2) / 2;
  }

  static DirectionMode detectMode(
    List<LatLng> ida,
    List<LatLng> vuelta, {
    double thresholdM = defaultMirrorThresholdM,
    String? propertyMode,
  }) {
    final m = (propertyMode ?? '').toLowerCase();
    if (m == 'mirrored') return DirectionMode.mirrored;
    if (m == 'independent') return DirectionMode.independent;
    if (m == 'dual_ring' || m == 'dualring') return DirectionMode.dualRing;

    final sim = mirrorSimilarityMeters(ida, vuelta);
    return sim <= thresholdM ? DirectionMode.mirrored : DirectionMode.independent;
  }

  /// Prepara shapes de explorador como `toSingleCorridorDisplay` de la web.
  ///
  /// - dual_ring / independent: ambas líneas reales ida+vuelta
  /// - mirrored: una geometría + vuelta como reverse para labels
  static List<RouteShapeModel> toCorridorDisplay(
    List<RouteShapeModel> shapes, {
    DirectionMode? forceMode,
    String preferDirection = 'both', // both | ida | vuelta
  }) {
    if (shapes.isEmpty) return const [];

    final byDir = <String, RouteShapeModel>{};
    for (final s in shapes) {
      byDir[s.direction] = s;
    }
    final ida = byDir['ida'];
    final vuelta = byDir['vuelta'];
    final propMode = ida?.directionMode ?? vuelta?.directionMode;
    final mode = forceMode ??
        detectMode(
          ida?.coordinates ?? const [],
          vuelta?.coordinates ?? const [],
          propertyMode: propMode,
        );

    if (preferDirection == 'ida' && ida != null) {
      return [ida.copyWith(role: 'full')];
    }
    if (preferDirection == 'vuelta' && vuelta != null) {
      return [vuelta.copyWith(role: 'full')];
    }
    if (preferDirection == 'ida' && ida == null && vuelta != null) {
      return [
        vuelta.copyWith(
          direction: 'ida',
          coordinates: reverseCoordinates(vuelta.coordinates),
          role: 'full',
        ),
      ];
    }
    if (preferDirection == 'vuelta' && vuelta == null && ida != null) {
      return [
        ida.copyWith(
          direction: 'vuelta',
          coordinates: reverseCoordinates(ida.coordinates),
          role: 'full',
        ),
      ];
    }

    // both
    if (mode == DirectionMode.mirrored) {
      final base = ida ?? vuelta;
      if (base == null) return shapes;
      final corridor = base.coordinates;
      final rev = reverseCoordinates(corridor);
      return [
        base.copyWith(
          id: '${base.routeId}-corridor',
          direction: 'ida',
          coordinates: corridor,
          directionMode: 'mirrored',
          role: 'full',
        ),
        base.copyWith(
          id: '${base.routeId}-sense-vuelta',
          direction: 'vuelta',
          coordinates: rev,
          directionMode: 'mirrored',
          role: 'sense-label',
        ),
      ];
    }

    // dual_ring / independent: ambas reales
    final out = <RouteShapeModel>[];
    if (ida != null) {
      out.add(ida.copyWith(directionMode: mode.name, role: 'full'));
    }
    if (vuelta != null) {
      out.add(vuelta.copyWith(directionMode: mode.name, role: 'full'));
    }
    if (out.isEmpty) return shapes.map((s) => s.copyWith(role: 'full')).toList();
    return out;
  }
}
